### PR TITLE
Reads balancer : adding tests to verify functionality of online commands

### PR DESCRIPTION
**Pull request by pdhiran**
1. Added positive and negative tests to verify the behaviour of new online command "pg-upmap-primary" and verify OSD change in the acting set.
2. Added test coverage for OSD add and OSD delete during PG splits.
Signed-off-by: Pawan Dhiran <pdhiran@redhat.com>

### CODE REVIEW

**Comment by pdhiran** [comment]
logs : 
1. http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-DYX59O/Verify_add_remove_osd_during_PG_split_0.log

File: suites/reef/rados/tier-2_rados_test-pg-split-merge.yaml [hunk 1]
```diff
@@ -93,6 +93,63 @@ tests:
...
+              pool_name: ecpool_1
+              pool_type: erasure
+              pg_num: 32
+              k: 4
+              m: 2
+              plugin: jerasure
+              rados_write_duration: 50
+              rados_read_duration: 30
+              byte_size: 256
+          - create_pool:
```
Comment (neha-gangadhar) [inline suites/reef/rados/tier-2_rados_test-pg-split-merge.yaml:136]: Small observation: Since we cannot verify on EC pool the functionalities here but only verifying the proper error message is thrown on EC pool operations may be we can have single EC pool? Any reason we have 2 EC pools created?

File: tests/rados/test_reads_balancer.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+                new_acting_set = rados_obj.get_pg_acting_set(pg_num=pgid)
+                if new_acting_set[0] != acting_set[1]:
+                    log.error(
+                        f"The acting set has not changed."
+                        f"Old AC : {acting_set}, New AC : {new_acting_set}"
+                    )
+                    raise Exception("AC not changed error")
+
+                time.sleep(5)
+                # Checking if it is possible to revert the changes made.from
```
Comment (neha-gangadhar) [inline tests/rados/test_reads_balancer.py:220]: Please check the comment sentence if anything needs to be updated here. May be "from" needs to be removed.

**Comment by neha-gangadhar** [comment]
LGTM with minor comments. Both tests are reviewed.

**Comment by pdhiran** [comment]
Complete suite pass logs : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-CZGX0V

File: suites/reef/rados/tier-2_rados_test-pg-split-merge.yaml (hunk 1 above)
Comment (pdhiran) [reply suites/reef/rados/tier-2_rados_test-pg-split-merge.yaml:136]: For the verification of online commands, we just do -ve scenario.

With the verification of offline tool, we want to make sure that applying reads balancing on RE pools does not cause adverse affect on reads of EC pools. hence 2 were created.

But yes, at this point, We can completely do tests with single EC pool as well.

Will do as suggested!

Comment (pdhiran) [reply suites/reef/rados/tier-2_rados_test-pg-split-merge.yaml]: Removed 1 EC pool creation.

File: tests/rados/test_reads_balancer.py (hunk 2 above)
Comment (pdhiran) [reply tests/rados/test_reads_balancer.py]: Corrected the comment

**Comment by pdhiran** [comment]
> LGTM with minor comments. Both tests are reviewed.

Review comments addressed and complete suite execution added for reef run.

**Comment by pdhiran** [comment]
Rebased with latest!

File: ceph/rados/core_workflows.py [hunk 3]
```diff
@@ -986,6 +986,16 @@ def verify_reweight(self, affected_osds: list, osd_info: list) -> bool:
 
         return True
 
+    def list_pools_on_clustr(self) -> list:
```
Comment (harshkumarRH) [inline ceph/rados/core_workflows.py:989]: Is this an intentional typo?

File: ceph/rados/core_workflows.py [hunk 4]
```diff
@@ -986,6 +986,16 @@ def verify_reweight(self, affected_osds: list, osd_info: list) -> bool:
...
+        """
+        Method fetches the list of pools from the cluster via ceph df command
+
+        Returns:
+            list of pools on the cluster
+        """
+        existing_pools = self.run_ceph_command(cmd="ceph df", client_exec=True)
+        return [ele["name"] for ele in existing_pools["pools"]]
```
Comment (harshkumarRH) [inline ceph/rados/core_workflows.py:997]: I believe `ceph osd pool ls -f json` would simplify the code
_**# ceph osd pool ls -f json
[".mgr","cephfs.cephfs.meta","cephfs.cephfs.data","pool_replica_less_min"]**_


`return self.run_ceph_command(cmd="ceph osd pool ls", client_exec=True)`

Let me know what you think

File: ceph/rados/core_workflows.py (hunk 3 above)
Comment (pdhiran) [reply ceph/rados/core_workflows.py:989]: No. Corrected the typo.

File: ceph/rados/core_workflows.py (hunk 4 above)
Comment (pdhiran) [reply ceph/rados/core_workflows.py:997]: Updated the code.

**Comment by pdhiran** [comment]
Review comments addressed. Please review the PR.

File: tests/rados/test_reads_balancer.py [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+            read_scores = pool_details["read_balance"]
+            if read_scores["score_acting"] > pool_details["size"]:
+                log.error(
+                    f"The balance score on the pool is more than the size of the pool."
+                    f"Balance score : {read_scores['score_acting']}, Size of poool : {pool_details['size']}"
+                )
+                log.error(
+                    "This failure will be ignored because of the bug "
+                    ": https://bugzilla.redhat.com/show_bug.cgi?id=2237352 and the failure will be added post fix"
+                )
```
Comment (harshkumarRH) [inline tests/rados/test_reads_balancer.py:74]: Maybe we should add an exception or return 1 in this error block as well

File: tests/rados/test_reads_balancer.py [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+            log.info(
+                f"Selected pool {pool_name} is a EC pool, skipping check of balance scores."
+            )
+        log.info("Completed verification of balance scores for all the pools")
+
+    log.debug(
+        "Setting config to allow clients to perform read balancing on the clusters."
+    )
+    config_cmd = "ceph osd set-require-min-compat-client reef"
+    rados_obj.run_ceph_command(cmd=config_cmd)
```
Comment (harshkumarRH) [inline tests/rados/test_reads_balancer.py:101]: Should the `ceph osd set....` command run with a `-f json` suffix? If not, maybe we can switch to `cephadm.shell([config_cmd])` instead of using `run_ceph_command`

File: tests/rados/test_reads_balancer.py [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+                    f"Error EINVAL: osd.{new_osd} is already primary for pg {pgid}"
+                )
+
+                try:
+                    out, err = client_node.exec_command(sudo=True, cmd=change_cmd)
+                except Exception as e:
+                    if error_str not in str(e).strip():
+                        log.error(
+                            "Correct error string not seen in stderr stream"
+                            f"Expected : {error_str} \n Got : {e}"
```
Comment (harshkumarRH) [inline tests/rados/test_reads_balancer.py:137]: f"Expected : {error_str} \n Got : {_**err**_}"

File: tests/rados/test_reads_balancer.py [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+                new_osd = acting_set[0]
+                change_cmd = f"ceph osd pg-upmap-primary {pgid} {new_osd}"
+                error_str = (
+                    f"Error EINVAL: osd.{new_osd} is already primary for pg {pgid}"
+                )
+
+                try:
+                    out, err = client_node.exec_command(sudo=True, cmd=change_cmd)
+                except Exception as e:
+                    if error_str not in str(e).strip():
```
Comment (harshkumarRH) [inline tests/rados/test_reads_balancer.py:134]: Should we be checking the error_str in STDERR or in the exception?
if error_str not in str(_**err**_).strip():

File: tests/rados/test_reads_balancer.py [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+                error_str = f"Error ENOENT: osd.{new_osd} does not exist"
+                try:
+                    out, err = client_node.exec_command(sudo=True, cmd=change_cmd)
+                except Exception as e:
+                    if error_str not in str(e).strip():
+                        log.error(
+                            "Correct error string not seen in stderr stream"
+                            f"Expected : {error_str} \n Got : {e}"
+                        )
+                        raise Exception("Wrong error string obtained")
```
Comment (harshkumarRH) [inline tests/rados/test_reads_balancer.py:157]: same comment as above

File: tests/rados/test_reads_balancer.py [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+                    log.error(f"An error occurred but was expected: {e}")
+                    log.debug(
+                        "Could not change the primary with non existent OSD. Pass"
+                    )
```
Comment (harshkumarRH) [inline tests/rados/test_reads_balancer.py:161]: As an enhancement later on, maybe we can modularize this section of the code, and provide the error string and command output as inputs

File: tests/rados/test_reads_balancer.py [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+                for node in ceph_cluster:
+                    if node.role == "osd":
+                        node_osds = rados_obj.collect_osd_daemon_ids(node)
+                        osd_list = osd_list + node_osds
+
+                pgid = f"{pool_id}.0"
+                acting_set = rados_obj.get_pg_acting_set(pg_num=pgid)
+                for item in osd_list:
+                    if item not in acting_set:
+                        new_osd = item
```
Comment (harshkumarRH) [inline tests/rados/test_reads_balancer.py:174]: maybe we should break from this for loop when we are getting an OSD which is not part of acting set, no point in continuing the loop as the purpose is fulfilled

File: tests/rados/test_reads_balancer.py [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+                try:
+                    out, err = client_node.exec_command(sudo=True, cmd=change_cmd)
+                except Exception as e:
+                    if error_str not in str(e).strip():
+                        log.error(
+                            "Correct error string not seen in stderr stream"
+                            f"Expected : {error_str} \n Got : {e}"
+                        )
+                        raise Exception("Wrong error string obtained")
+                    log.error(f"An error occurred but was expected: {e}")
```
Comment (harshkumarRH) [inline tests/rados/test_reads_balancer.py:189]: same as above

File: tests/rados/test_reads_balancer.py [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+                # Trying to change the primary with secondary. Should be possible
+                pgid = f"{pool_id}.0"
+                acting_set = rados_obj.get_pg_acting_set(pg_num=pgid)
+                new_osd = acting_set[1]
+                change_cmd = f"ceph osd pg-upmap-primary {pgid} {new_osd}"
+                error_str = f"change primary for pg {pgid} to osd.{new_osd}"
+                out, err = client_node.exec_command(sudo=True, cmd=change_cmd)
+                if error_str not in err.strip():
+                    log.error(
+                        "Correct error string not seen in stderr stream"
```
Comment (harshkumarRH) [inline tests/rados/test_reads_balancer.py:203]: Maybe rephrase it to "Correct pass string not seen in STDERR"

File: tests/rados/test_reads_balancer.py [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+                        f"Expected : {error_str} \n Got : {err}"
+                    )
+                    raise Exception("wrong error string obtained")
```
Comment (harshkumarRH) [inline tests/rados/test_reads_balancer.py:206]: As this is a positive scenario, maybe change the Exception msg to - "Expected pass string not found"

File: tests/rados/test_reads_balancer.py [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+                final_acting_set = rados_obj.get_pg_acting_set(pg_num=pgid)
+                if final_acting_set[0] != acting_set[0]:
+                    log.error(
+                        f"The acting set has not reverted to original."
+                        f"Old AC : {acting_set}, final AC : {final_acting_set}"
+                    )
+                    raise Exception("AC not changed error")
+
+                log.info("Completed tests on replicated pools for online commands")
+                break
```
Comment (harshkumarRH) [inline tests/rados/test_reads_balancer.py:234]: If we break here, we would be checking this workflow for only one replicated pool, I hope that's the intention.
Also if that's the intention, maybe instead of going through a _for_ loop, we can choose the first replicated pool we created and run the whole _online_command_verification_ workflow for that pool itself

File: tests/rados/test_reads_balancer.py [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+            # Balance_score = raw_score / Optimal_score
+            expected_score = (
+                read_scores["raw_score_acting"] / read_scores["optimal_score"]
+            )
+            if round(read_scores["score_acting"], 2) != round(expected_score, 2):
+                log.error(
+                    f"The scores for the pool is not calculated properly. acting score {read_scores['score_acting']}"
+                    f"Expected : {(read_scores['raw_score_acting'] / read_scores['optimal_score'])}"
+                )
+                raise Exception("Scores out of bound failure")
```
Comment (harshkumarRH) [inline tests/rados/test_reads_balancer.py:89]: I'm unable to find a `try, except, finally` block to catch this exception. Please share the line# if I'm wrong

File: tests/rados/test_reads_balancer.py (hunk 5 above)
Comment (pdhiran) [reply tests/rados/test_reads_balancer.py:74]: It's present, but it's commented out. The issue not not blocker for test to fail.

File: tests/rados/test_reads_balancer.py (hunk 6 above)
Comment (pdhiran) [reply tests/rados/test_reads_balancer.py:101]: It does not matter either way. there is no loss or gain by doing that.

File: tests/rados/test_reads_balancer.py (hunk 7 above)
Comment (pdhiran) [reply tests/rados/test_reads_balancer.py:137]: Both are same. does not matter either way.

File: tests/rados/test_reads_balancer.py (hunk 8 above)
Comment (pdhiran) [reply tests/rados/test_reads_balancer.py:134]: As soon as the command fails, we land in exception. and both "err" & "e" are same. Does not matter either way.

File: tests/rados/test_reads_balancer.py (hunk 10 above)
Comment (pdhiran) [reply tests/rados/test_reads_balancer.py:161]: Noted.

File: tests/rados/test_reads_balancer.py (hunk 11 above)
Comment (pdhiran) [reply tests/rados/test_reads_balancer.py:174]: Done.

File: tests/rados/test_reads_balancer.py (hunk 13 above)
Comment (pdhiran) [reply tests/rados/test_reads_balancer.py:203]: Done.

File: tests/rados/test_reads_balancer.py (hunk 14 above)
Comment (pdhiran) [reply tests/rados/test_reads_balancer.py:206]: Done.

File: tests/rados/test_reads_balancer.py (hunk 15 above)
Comment (pdhiran) [reply tests/rados/test_reads_balancer.py:234]: Just kept it like that so that in future, after completing the entire workflow of the reads balancer, i can run it on all the pools.

I can remove if needed, but i'm inclined to keep it.

File: tests/rados/test_reads_balancer.py (hunk 16 above)
Comment (pdhiran) [reply tests/rados/test_reads_balancer.py]: I did not have one in the test. added.

File: tests/rados/test_reads_balancer.py [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+                            log.error(
+                                "Correct error string not seen in stderr stream"
+                                f"Expected : {error_str} \n Got : {e}"
+                            )
+                            raise Exception("Wrong error string obtained")
+                        log.error(f"An error occurred but was expected: {e}")
+                        log.debug("Could not change the primary for EC pools. Pass")
+                    break
+                log.info("Verified the Online commands to change acting set")
+                return 0
```
Comment (harshkumarRH) [inline tests/rados/test_reads_balancer.py:282]: Very minor, these two lines should be outside the for loop starting at line #248

**Review by harshkumarRH — APPROVED** [review]
LGTM

**Comment by pdhiran** [comment]
Pass logs : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-PP76NK

**Review by neha-gangadhar — APPROVED** [review]
